glogger/handler.py

```diff
@@ -0,0 +1,277 @@
+        # Leave a KB buffer for all the short items + any overhead:
+        max_field_length = max(minimum, (self.max_message_size - 1024) // len(long_items))
+        for k, v in long_items:
+            d[k] = v[:max_field_length]
+        d["truncated"] = True
+
+    def warn_dropped(self, dropped: int) -> None:
+        # If we use logging, we risk getting here more than once. No can do.
+        self.messages_buffer.append(
+            self.jsonify(
+                {
+                    "serial_no": self.messages_buffer.next_serial_no,
+                    "timestamp": datetime.utcnow().isoformat(),
+                    "logger_name": "glogger",
+                    "severity": logging.WARNING,
+                    "message": f"Maximum total length ({self.messages_buffer.max_total_length}) exceeded. "
+                    f"Dropped {dropped} messages.",
+                }
+            )
+        )
+
+    def _flush_loop(self) -> None:
+        self.last_flush_time = self.time_fn()
+        while not self.stop_event.is_set():
+            if self.should_flush():
+                self._flush()
+            self.stop_event.wait(0.5)
+        # Flush all remaining messages before terminating:
+        if self.messages_buffer.count > 0:
+            self._flush()
+
+    @property
+    def time_since_last_flush(self) -> float:
+        return self.time_fn() - self.last_flush_time
+
+    def should_flush(self) -> bool:
+        return (
+            self.messages_buffer.count > 0
+            and (self.messages_buffer.utilized >= self.flush_threshold)
+            or (self.time_since_last_flush >= self.flush_interval)
+        )
+
+    # This is deliberately not "flush", because logging.shutdown() calls flush() and we don't want
+    # any flushing to happen from multiple threads.
+    def _flush(self) -> None:
+        # Allow configuring max_tries via class member. Alternatively we could decorate _send_logs in __init__ but that
+        # would be easier to miss and less flexible.
+        send = backoff.on_exception(backoff.expo, exception=RequestException, max_tries=self.max_send_tries)(
+            self._send_logs
+        )
+
+        self.last_flush_time = self.time_fn()
+        try:
+            batch, response = send()
+            response.raise_for_status()
+        except Exception:
+            logger.exception(SERVER_SEND_ERROR_MESSAGE)
+        else:
+            self._drop_sent_batch(batch)
+
+    def _send_logs(self) -> Tuple[Batch, requests.Response]:
+        """
+        :return: (batch that was sent, response)
+        """
+        # Upon every retry we will remake the batch, in case we are able to batch more messages together.
+        batch = self.make_batch()
+        batch_data = {
+            "batch_id": batch.ident,
+            "metadata": self.get_metadata(),
+            "lost": batch.lost,
+            "logs": "<LOGS_JSON>",
```

Comment:
I really don't like this, if we're trying to optimize let's add a level of cache so we won't do the json() twice.